openrtb_ext/convert_up_test.go

```diff
@@ -0,0 +1,382 @@
+			givenRequest: openrtb2.BidRequest{},
+		},
+		{
+			description:  "Ext",
+			givenRequest: openrtb2.BidRequest{Ext: json.RawMessage("malformed")},
+			expectedErr:  "req.ext is invalid: invalid character 'm' looking for beginning of value",
+		},
+		{
+			description:  "Source.Ext",
+			givenRequest: openrtb2.BidRequest{Source: &openrtb2.Source{Ext: json.RawMessage("malformed")}},
+			expectedErr:  "req.source.ext is invalid: invalid character 'm' looking for beginning of value",
+		},
+		{
+			description:  "Regs.Ext",
+			givenRequest: openrtb2.BidRequest{Regs: &openrtb2.Regs{Ext: json.RawMessage("malformed")}},
+			expectedErr:  "req.regs.ext is invalid: invalid character 'm' looking for beginning of value",
+		},
+		{
+			description:  "User.Ext",
+			givenRequest: openrtb2.BidRequest{User: &openrtb2.User{Ext: json.RawMessage("malformed")}},
+			expectedErr:  "req.user.ext is invalid: invalid character 'm' looking for beginning of value",
+		},
+	}
+
+	for _, test := range testCases {
+		w := &RequestWrapper{BidRequest: &test.givenRequest}
+		err := convertUpEnsureExt(w)
+		if len(test.expectedErr) > 0 {
+			assert.EqualError(t, err, test.expectedErr, test.description)
+		} else {
+			assert.NoError(t, err, test.description)
+		}
+	}
+}
+
+func TestMoveSupplyChainFrom24To25(t *testing.T) {
+	var (
+		schain1 = json.RawMessage(`{"schain":{"complete":1,"nodes":[],"ver":"1"}}`)
+		schain2 = json.RawMessage(`{"schain":{"complete":1,"nodes":[],"ver":"2"}}`)
+	)
+
+	testCases := []struct {
+		description     string
+		givenRequest    openrtb2.BidRequest
+		expectedRequest openrtb2.BidRequest
+	}{
+		{
+			description:     "Not Present",
+			givenRequest:    openrtb2.BidRequest{},
+			expectedRequest: openrtb2.BidRequest{},
+		},
+		{
+			description:     "2.4 Migrated To 2.5 - Source Doesn't Exit",
+			givenRequest:    openrtb2.BidRequest{Ext: schain1},
+			expectedRequest: openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}},
+		},
+		{
+			description:     "2.4 Migrated To 2.5 - Source Exits",
+			givenRequest:    openrtb2.BidRequest{Source: &openrtb2.Source{}, Ext: schain1},
+			expectedRequest: openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}},
+		},
+		{
+			description:     "2.4 Dropped",
+			givenRequest:    openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}, Ext: schain2},
+			expectedRequest: openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}},
+		},
+		{
+			description:     "2.5 Left Alone",
+			givenRequest:    openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}},
+			expectedRequest: openrtb2.BidRequest{Source: &openrtb2.Source{Ext: schain1}},
+		},
+	}
+
+	for _, test := range testCases {
+		w := &RequestWrapper{BidRequest: &test.givenRequest}
+		moveSupplyChainFrom24To25(w)
+		assert.NoError(t, w.RebuildRequest(), test.description)
+		assert.Equal(t, test.expectedRequest, *w.BidRequest, test.description)
+	}
+}
+
+func TestConvertSupplyChainFrom25To26(t *testing.T) {
```

Comment:
Should we add another test case here to cover the scenario where the source object has to be created before moving the 2.5 schain to the 2.6 location? You already have a test like this for moving from 2.4 to 2.5.

Reply:
I prefer not to cover that case in the tests for `moveSupplyChainFrom25To26` and keep coverage in `TestConvertUpTo26` because this is an emergent behavior combining the two together due to the nuances of the request wrapper. I think an integration test is the most appropriate location for that coverage. Cool with you?

Reply:
Good point, cool with me.